### ISSUE TITLE
Calendar import of ics is not working

### ISSUE THREAD

**Issue by arichiardi** [xwiki-labs/cryptpad#784]
**Describe the bug**
Nothing is imported when I try to import an `ics` file.

**Where did it happen?**
CryptPad.fr

**To Reproduce**
1. Go to Calendar
2. Create a New Calendar
3. Click on the `...`
4. Click on Import
4. Select a local ics file

**Expected behavior**
I would like to see the events imported in my new calendar

**Browser (please complete the following information):**
 - OS: linux
 - Firefox
 - 91.0 
 - Only GetPocket
 - No major tweaks

**Additional context**
I can see this desplayed in the Dev Tools Console:

```
Uncaught TypeError: end is null
    import https://cryptpad.fr/calendar/export.js?ver=4.10.0-00:150
    import https://cryptpad.fr/calendar/export.js?ver=4.10.0-00:142
    execCb https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1696
    check https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:883
    enable https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1176
    init https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:788
    localRequire https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1460
    setTimeout handler*req.nextTick< https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1815
    localRequire https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1449
    requirejs https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1797
    import https://cryptpad.fr/calendar/export.js?ver=4.10.0-00:117
    action https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:473
    onload https://cryptpad.fr/common/common-ui-elements.js?ver=4.10.0-00:142
    importContent https://cryptpad.fr/common/common-ui-elements.js?ver=4.10.0-00:142
    jQuery 8
    importContent https://cryptpad.fr/common/common-ui-elements.js?ver=4.10.0-00:137
    action https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:486
    createDropdown https://cryptpad.fr/common/common-ui-elements.js?ver=4.10.0-00:1447
    jQuery 9
    createDropdown https://cryptpad.fr/common/common-ui-elements.js?ver=4.10.0-00:1446
    createDropdown https://cryptpad.fr/common/common-ui-elements.js?ver=4.10.0-00:1440
    makeEditDropdown https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:602
    makeCalendarEntry https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:610
    makeLeftside https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:706
    makeLeftside https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:705
    fire https://cryptpad.fr/common/common-util.js?ver=4.10.0-00:85
    fire https://cryptpad.fr/common/common-util.js?ver=4.10.0-00:85
    onCalendarUpdate https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:212
    onEvent https://cryptpad.fr/calendar/inner.js?ver=4.10.0-00:1100
    fire https://cryptpad.fr/common/common-util.js?ver=4.10.0-00:85
    fire https://cryptpad.fr/common/common-util.js?ver=4.10.0-00:85
    create https://cryptpad.fr/common/sframe-common.js?ver=4.10.0-00:850
    h https://cryptpad.fr/common/outer/worker-channel.js?ver=4.10.0-00:96
    create https://cryptpad.fr/common/outer/worker-channel.js?ver=4.10.0-00:175
    create https://cryptpad.fr/common/outer/worker-channel.js?ver=4.10.0-00:174
    fire https://cryptpad.fr/common/common-util.js?ver=4.10.0-00:85
export.js:150:21
```

**Comment by ansuz** [comment]
Thanks for the detailed report!

We use [Mozilla Thunderbird's ical.js](https://github.com/mozilla-comm/ical.js) to parse ICS files. It looks like the set of values returned by this library when parsing your ICS file did not include an `DTEND` field, which we were expecting.

From reading specifications and [some commentary on the ICS format](https://stackoverflow.com/questions/15295887/does-an-ical-file-have-to-have-an-enddate/15308766) it seems like one of two things should happen when no DTEND field is present depending on the context of the other fields in the file:

1. if DTSTART is a day, DTEND is one day later
2. if DTSTART is a time, DTEND is strictly equal to DTSTART (implying that the event has a duration of zero seconds)

We've written [a small patch](https://github.com/xwiki-labs/cryptpad/commit/904e06091da39b6f56ef88382091e2411a30ff0f) which checks and adjusts the returned values if the ones we need are not present. This will be included in our next release.

It might be helpful if you could share the content of the ICS file which triggered this error. If the details of the event are sensitive an alternative is to try and create a new ICS file in the same manner without those sensitive details and confirm that it produces the same result.

If neither option is possible, there are two online validators that can be used to check the validity of the file:

* [one provided by icalendar.org](https://icalendar.org/validator.html#results)
* [one which uses ical.js directly](https://mozilla-comm.github.io/ical.js/validator.html)

In any case, it would be interesting to know whether these two agree about the validity of your ICS file.

**Comment by arichiardi** [comment]
Thank *you* for the details!

I have tried the attached file in both validators and the former throws a couple of errors related to `DSTAMP`

```
Errors

    Missing DTSTAMP property near line # 9Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 21Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 33Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 45Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 57Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 69Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 81Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 93Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 105Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 117Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 129Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 141Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 153Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 165Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 177Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 189Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 201Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 213Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 225Reference: RFC 5545 3.6.1. Event Component
    Missing DTSTAMP property near line # 237Reference: RFC 5545 3.6.1. Event Component
```

No mention of `DTEND`. The latter is actually parsing it correctly.

As another data point, this calendar has got a shorter error in the console compared to the one I tried first:

```
Uncaught TypeError: end is null
    import https://cryptpad.fr/calendar/export.js?ver=4.10.0-00:150
    import https://cryptpad.fr/calendar/export.js?ver=4.10.0-00:142
    execCb https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1696
    check https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:883
    enable https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1139
    bind https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:134
    emit https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1189
    each https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:59
    emit https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1188
    check https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:938
    enable https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1176
    init https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:788
    callGetModule https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1203
    completeLoad https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1611
    onScriptLoad https://sandbox.cryptpad.info/bower_components/requirejs/require.js?ver=2.3.5:1717
export.js:150:21
```

Cannot wait for the patch at this point :smile: Thanks for addressing it so quickly!

**Comment by ansuz** [comment]
This was included in our 4.11.0 release.
